AD Helpers: Introduce base class for AD helper classes

jppelteret opened the pull request:
Blocked by #5499

**TODO**:
- [x] Update documentation (headers)
- [x] Update documentation (source)
- [x] Check validity of discussions wrt. both `ADOL-C` and `Sacado`
- [x] Proofread

jppelteret commented:
@bangerth Rebased :-)

jppelteret commented:
/run-tests

bangerth commented:
Oh, another 1500 lines of code...

jppelteret commented:
> Oh, another 1500 lines of code...

In my defence, its one class and many lines of comments :-(

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+     * A base helper class that facilitates the evaluation of the derivative(s)
+     * of a number of user-defined dependent variables $\mathbf{f}(\mathbf{X})$
+     * with respect to a set of independent variables $\mathbf{X}$, that is
+     * $\dfrac{d^{i} \mathbf{f}(\mathbf{X})}{d^{i} \mathbf{X}}$.
```
in the denominator:
d^{i} \mathbf{X}  -> d \mathbf{X}^{i}

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+     *   // Extract the sensitive equivalent of the independent variables
+     *   // The operations conducted with these numbers will be tracked.
+     *   const auto ad_independent_variables
+     *     = ad_helper.get_sensitive_variables(...);
```
how does this work? how would the `ad_helper` know which of the independent variables are sensitive and which are not?

also, I wonder whether there is a way to define the term "sensitive". Maybe a glossary entry, and a link to it from the first place here (and maybe elsewhere) where you use it?

jppelteret replied:
By definition, all of the independent variables will be made "sensitive". Internally, we create a vector of AD numbers that are _marked_ as being independent variables. The values sent into the `register_independent_variable()` function above are used to initialise these AD counterparts. I've made a small note to this effect in the code examples.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+     * in the individual classes that specialize this base class.
+     *
+     * When "taped" numbers are to be used, the above code should be wrapped by
+     * a few more lines of code to manage the taping procedure:
```
I'm not enough of an expert to define what "taped" actually means. Maybe that too could become a glossary entry? My intuition is that you can use a tape to record the operations done only once, and so don't have to compute all of the derivatives every time you enter a function, but can reuse what's been done in a previous invokation. Is this about right?

jppelteret replied:
Yes, lets add glossary entries for these. But maybe I could do it as a follow up PR (see https://github.com/dealii/dealii/issues/7120).

jppelteret replied:
BTW. I think I give a pretty detailed description of "taped" in the [AD module](https://dealii.org/developer/doxygen/deal.II/group__auto__symb__diff.html). But this should still be added to a glossary entry.

bangerth replied:
Separate PR is totally fine.

If you want to, you can also just link to that module -- maybe that's the better place anyway, since it is the one central place where everything AD related comes together. Either way is fine by me, though.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+     *     = ad_helper.enable_record_sensitivities(tape_index);
+     *   if (is_recording == true)
+     *   {
+     *     // In this block on places the majority of the operations described
```
on -> one

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+     * numbers.
+     *
+     * @warning ADOL-C does not support the standard threading models used by
+     * deal.II, so this class should \b not be embedded within a multithreaded
```
@b

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+    {
+    public:
+      /**
+       * Type definition for the floating point number type that are used in,
```
are -> is

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * desired to compute the sensitivities of.
+       * @param[in] n_dependent_variables Number of scalar functions to be
+       * defined that will have a sensitivity to the given independent
+       * variables.
```
is this a *subset of*, or *in addition to* the independent variables?

also, is it correct that these are the "sensitive" variables?

jppelteret replied:
I've now added some further text to clarify your misunderstanding here. Considering `\mathbf{f}(\mathbf{X})`, this is the number of outputs `\mathbf{f}`.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * work with.
+       */
+      std::size_t
+      n_independent_variables() const;
```
maybe say that this is the dimension of the range space

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * operate on.
+       */
+      std::size_t
+      n_dependent_variables() const;
```
maybe say that this is the dimension of the image space

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * written.
+       */
+      void
+      print(std::ostream &stream) const;
```
can you say that what is actually being printed depends on the AD type used (and specified by the template argument)? At least that's what I think is happening, not having read down to the implementation of the function.

Is this function overloaded in derived classes? If so, should it be `virtual`?

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       */
+      void
+      print_tape_stats(std::ostream &           stream,
+                       const types::tape_index &tape_index) const;
```
revert the arguments to stick with our usual "input arguments first" approach

bangerth replied:
the index also probably doesn't have to be a reference

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * @warning With @p ensure_persistent_setting set to <tt>true</tt>,
+       * calling this function leaves the set number of directional
+       * derivatives in a persistent state, so it will not be possible to
+       * further modify during course of the program's execution.
```
modify WHAT?

jppelteret replied:
I've extended the documentation to clarify this point.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * further modify during course of the program's execution.
+       */
+      static void
+      configure_tapeless_mode(const unsigned int &n_independent_variables,
```
no need for the reference

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * it to have any influence on the memory allocated to the next recorded
+       * tape.
+       *
+       * @note This only has an effect when using ADOL-C numbers. As stated by
```
this -> this function

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * @param[in] tbufsize ADOL-C Taylor buffer size
+       */
+      void
+      set_tape_buffer_sizes(const types::tape_buffer_sizes obufsize = 67108864,
```
where do these defaults come from? might be useful to explain the default values in the documentation above

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * of disable_record_sensitivities() are recorded to the tape and can
+       * be replayed and reevaluated as necessary.
+       *
+       * The typical set of operations to be performed during this phase are:
```
with "this phase", do you mean "the phase between enable_...() and disable_...() calls"? If so maybe make this clearer. For example, you could give this phase a name in the previous paragraph and reference it here.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * @p tape_index will be destroyed and a new tape traced over it.
+       * @param[in] keep_values Determines whether the numerical values of all
+       * independent variables are recorded in the tape buffer. If true, then
+       * tape can be immediately used to perform computations after recording
```
then tape -> then the tape

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * types::invalid_tape_index and less than types::max_tape_index.
+       */
+      bool
+      enable_record_sensitivities(const types::tape_index &tape_index,
```
Have you thought about calling these functions `start_record_sensitivities()` and `stop_record_sensitivities()`? That's because a strict interpretation of the word would imply that `enable/disable` really only says that you make it possible/impossible to record, but not necessarily that it actually happens. 

I don't know whether it's necessary to make this change, but wanted to bring it up anyway.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+      /**
+       * Select a pre-recorded tape to read from.
+       *
+       * @param[in] tape_index The index of the tape to be read from
```
period missing at the end

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * @param[in] tape_index The index of the tape to be read from
+       *
+       * @note The chosen tape index must be greater than
+       * types::invalid_tape_index and less than types::max_tape_index.
```
oh, are these numbers defined in namespace `types::`? we usually define these special numbers in namespace `numbers::` (e.g., `numbers::invalid_dof_index`)

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * types::invalid_tape_index and less than types::max_tape_index.
+       */
+      void
+      activate_recorded_tape(const types::tape_index &tape_index);
```
no need for the reference

jppelteret commented:
BTW. I'm not ignoring these comments. I'm just waiting for a full review to come in before making changes.

bangerth commented:
Yes, this got away from me while on vacation last week. I'll get back to it.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * The set of dependent variables $\mathbf{f}(\mathbf{X})$ of which the
+       * derivatives with respect to $\mathbf{X}$ will be computed.
+       *
+       * The gradients and Hessians of these dependent variables will be
```
Somewhere higher up in this file, I saw you spell "hessian" in lower-case letters. I prefer "Hessian" and "Jacobian". In the end, that's not important, but while you're editing around here, choose a common spelling.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+      /**
+       * A list of dependent variables.
+       */
+      std::vector<bool> touched_dependent_variables;
```
A list of *touched* dependent variables?

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+
+      /**
+       * Reset the boolean vector that indicates which independent variables
+       * that have been manipulated by the current set of operations.
```
is the "that" spurious here? This looks grammatically odd to me.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+      /**
+       * Reset the boolean vector that indicates which independent variables
+       * that have been manipulated by the current set of operations.
+       * All entries in the vector are set to the value of the @p flag.
```
can you say which vector this is? This is an internal function, so it makes sense to reference the internal variable in question here

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * All entries in the vector are set to the value of the @p flag.
+       */
+      void
+      reset_touched_dependent_variables(const bool flag = false);
```
would you ever want to call this function with `true` as argument?

jppelteret replied:
Yes, it is necessary in one location. See [here](https://github.com/dealii/dealii/pull/6975/files#diff-c8b253323ba2c5c505cc06c2a6138d07R568) for context.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * The number of dependent variables that have been registered.
+       */
+      unsigned int
+      n_touched_dependent_variables() const;
```
I get confused between the terms "touched" and "registered". Is there a difference? If not, maybe settle on one of the terms?

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       */
+      void
+      register_dependent_variable(const ad_type &    func,
+                                  const unsigned int index);
```
maybe same here for the order of arguments as mentioned previously.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * of these helpers. However, it is assumed that when all helpers go
+       * out of scope then the tapes can be written over.
+       */
+      static dealii::types::global_dof_index n_helpers;
```
is `global_dof_index` the right type here? I would think that a regular `unsigned int` is enough, or alternatively a `std::uint64_t`.

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    dealii::types::global_dof_index
+      ADHelperBase<dim, ADNumberTypeCode, ScalarType>::n_helpers = 0;
```
this is the place where you use the type mentioned in the previous comment above in the definition of the variable

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+      // initializer.
+      dependent_variables =
+        std::vector<ad_type>(n_dependent_variables,
+                             dealii::internal::NumberType<ad_type>::value(0.0));
```
you achieve the same by saying
```
  dependent_variables.resize (n_dependent_variables, 
                                                dealii::...::value(0.0));
```

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+          Assert(this->is_recording() == true,
+                 ExcMessage(
+                   "Cannot change the value of an independent variable "
+                   "of the tapeless variety."));
```
do you want to say "...while recording" here?

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+          Assert(active_tape() != types::invalid_tape_index,
+                 ExcMessage("Invalid tape index"));
+          Assert(is_recording() == true,
+                 ExcMessage("Only valid during recording"));
```
can you say *what* is only valid during recording in the error message?

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+      if (this->independent_variables.size() == 0)
+        {
+          this->independent_variables =
+            std::vector<ad_type>(this->n_independent_variables(),
```
like above, you could use `.resize` here.

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+                 ExcMessage("Invalid tape index"));
+        }
+      Assert(is_recording() == false,
+             ExcMessage("Only valid outside of recording"));
```
as above

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+    {
+      return std::accumulate(touched_independent_variables.begin(),
+                             touched_independent_variables.end(),
+                             0u);
```
this is hard to understand. it works because the data type of the array (`bool`) can be converted to `unsigned int`, but I think what you really want to do is `std::count (...begin(), ...end(), true)`

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+              typename ScalarType>
+    std::size_t
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::n_independent_variables()
+      const
```
indentation fail :-(

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+    {
+      return std::accumulate(touched_dependent_variables.begin(),
+                             touched_dependent_variables.end(),
+                             0u);
```
same as above

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+      stream << "Active tape index: " << active_tape_index << "\n";
+      stream << "Registered tapes: ";
+      for (unsigned int i = 0; i < registered_tapes.size(); ++i)
+        stream << *(std::next(registered_tapes.begin(), i))
```
That's an `n^2` algorithm here. I suspect that that doesn't matter because we will never have a significant number of tapes registered, right?

jppelteret replied:
Thanks for pointing this out. Yes, there should not be many tapes in existence. But I've fixed this nonetheless.

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+        stream << *(std::next(registered_tapes.begin(), i))
+               << (i < (registered_tapes.size() - 1) ? "," : "");
+      stream << "\n";
+      stream << "Recording? " << is_recording() << "\n";
```
here and below you are outputting bool values, but by default C++ prints them as 1 and 0. If you wanted "true" and "false", you can do this by using `std::boolalpha` (https://en.cppreference.com/w/cpp/io/manip/boolalpha), but you would have to save/restore the stream state for that around this whole block.

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+             << "\n";
+      for (unsigned int i = 0; i < n_dependent_variables(); i++)
+        stream << touched_dependent_variables[i]
+               << (i < (n_dependent_variables() - 1) ? "," : "") << std::flush;
```
we really ought to invent a function that takes a container and outputs it with separators, rather than having to repeat this whole block over and over.

no need to do this here and now, but maybe keep it in mind for the future.

bangerth replied:
Separately, the `std::flush` is really not very useful here -- you later push `std::endl` into the stream, which implies a flush. The only reason you may want to do the flush here is if you anticipate that any of the following operations triggers an `abort` and you're afraid that you won't make it to the `std::endl` -- though I can't really see how that would happen here.

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+    {
+      const unsigned int new_n_independent_variables =
+        (n_independent_variables != 0 ? n_independent_variables :
+                                        this->n_independent_variables());
```
I didn't pay attention when I read the function docs, but are you using zero as a special value here? I think that's not a great choice. Why not use `numbers::invalid_unsigned_int` as everywhere else?

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -0,0 +1,783 @@
+      const types::tape_buffer_sizes in_obufsize,
+      const types::tape_buffer_sizes in_lbufsize,
+      const types::tape_buffer_sizes in_vbufsize,
+      const types::tape_buffer_sizes in_tbufsize)
```
this `types` namespace is inside namespace `AD`, right? I ask, because in many other contexts here, you explicitly qualify with `AD::`.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperBase
```
I can't seem to figure out where `dim` is actually used in these classes. Can you elaborate?

bangerth commented:
By the way, massive amount of work! I particularly appreciate the extensive documentation and error checking in this code -- great job!

bangerth commented:
@jppelteret -- let me know if you want me to take another look.

I suspect that the other pull requests you had posted are queued behind this one. Or are they independent?

jppelteret commented:
Thanks @bangerth for going through all of this so thoroughly. I'm sorry that I've not been able to get stuck into your comments yet. I guess I'll only have time from Friday onwards to work though all of them.

I had intended the PR's to be worked through sequentially, so apart from the fact that this is the base class referenced in #6976 its pretty much independent. So I guess you could continue and review `ADHelperCellLevelBase` in #6976 (followed by `ADHelperVariationalFormulation` in #6977 and `ADHelperResidualLinearisation` in #6978, if you feel so inclined).

bangerth commented:
OK, there's no rush from my side. I just wanted to make sure I'm not even more of a roadblock than I already am in all of this!

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+      /**
+       * Select a tape to record to or read from.
+       *
+       * @param[in] tape_index The index of the tape to be written / read from.
```
> I imagine typographists would point out that / should not have surrounding spaces, but I'll leave this up to you.

Looks gross, but I've fixed this up.

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -0,0 +1,683 @@
+       * @param[in] index The index in the vector of independent variables.
+       */
+      void
+      mark_independent_variable(ad_type &out, const unsigned int index) const;
```
> but why not just return the ad_type thing

There are two reasons for this that I can think of off the top of my head:
1. We want to minimise the number of active variables that we generate before handing things off to the user, so each pass by copy creates some overhead. In this way we can avoid that.
2. I've had some instances of data corruption that I've since avoided by never using a pass by copy. I can't remember how these issues came to be and how pervasive the issue was, but it was so frustrating to track down the issue that I simply made a rule to avoid copying AD numbers.

bangerth replied:
Hm, ok. It's not inconsistent with what we do with vectors, for example, just annoying that we can't do copies with what one would intuitively consider scalars.

jppelteret replied:
Well, since this is an internal function it doesn't really matter too much what the interface is. Its better that we pass a minimally polluted set of independent variables (the starting point of all AD computations) to our users. With the benchmarking I've done I get the sense that each superfluous operation adds a non-insignificant overhead to the derivative computations performed later.

bangerth replied:
Fair enough. I had also forgotten that it's an internal interface.

jppelteret commented:
Thanks for your patience with me @bangerth. Hopefully the other PR's will be less work to review since they heavily leverage what we've laid out here.

I think that I incorporated all of your suggestions and fixed all of the problematic things that you pointed out. All of my unit tests pass with the modifications applied.

jppelteret commented:
> A separate question: You use mutable and static member variables. This is generally a good way to get me thinking about multithreaded contexts, because these variables then present opportunities for race conditions. Have you thought about what happens if one uses these helper classes in such contexts, most notably when doing assembly of linear systems on multiple threads at the same time? Is this supported at all by the underlying libraries, and if so what would one need to do to use this?

I must admit that since ADOL-C doesn't support multithreading other than by OpenMP (for its taped type, at the very least), I completely ignored structuring this around the idea of multithreaded applications. I'll create an issue and will look into this in the future, since I suppose Sacado and maybe ADOL-C tapeless types might support it.

bangerth commented:
OK, so you think this is ready for merge now?

bangerth commented:
As for multithreading: I have no real idea whether that works with the other libraries. But it's a pretty common idiom in deal.II now, so I thought it was worth asking how that would fit together.

jppelteret commented:
> OK, so you think this is ready for merge now?

Yes, I think so.

> As for multithreading: I have no real idea whether that works with the other libraries. But it's a pretty common idiom in deal.II now, so I thought it was worth asking how that would fit together.

As I said, I'll investigate this before the release but I'd rather move forward with the rest of the helper classes now that we have some momentum with this.

bangerth commented:
Here we go. Please rebase all of the other PRs and let me know what to review next!
